[addons] next binary addon gui interface step

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This add the minimum parts to create a window on addon. This still
mostly equal with the Python addon window parts.

During test has this worked, but my experience for GUI things is not the
best and want a feedback.



## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented on xbmc/addons/interfaces/AddonInterfaces.cpp:
```diff
@@ -226,6 +227,9 @@ void CAddonInterfaces::OnApplicationMessage(ThreadMessage* pMsg)
       {
+      case 0:
+        static_cast<ADDON::CGUIAddonWindowDialog*>(pMsg->lpVoid)->Show_Internal(pMsg->param2 > 0);
+        break;
```
This present for downward compatibility, on end this becomes complete removed.

AlwinEsch commented on xbmc/addons/interfaces/GUI/ListItem.cpp:
```diff
@@ -0,0 +1,94 @@
+  // @note: Delete of CFileItem brings crash, need to check about the in related
+  // window and no memleak present to confirm.
+  // In old version was the same way, only the Destroy passed here to allow
+  // changes and fixes without API level changes.
```
The list item is currently not used, but added. Want to check on next steps this note and to make really sure the list items becomes deleted.

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -0,0 +1,657 @@
+{
+  Interface_GUIGeneral::lock();
+
+  // window id's 14000 - 14100 are reserved for addons
```
I would remove the hard numbers here and in the following lines as we have WINDOW_ADDON_START and WINDOW_ADDON_END

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/ListItem.h:
```diff
@@ -0,0 +1,118 @@
+    CListItem(GUIHANDLE listItemHandle)
+      : CAddonGUIControlBase(nullptr)
+    {
+      m_controlHandle = listItemHandle;
```
This might be the cause of the crashes you experience when <code>Interface_GUIListItem::destroy</code> is called. The original and the copy share a pointer to a single CFileItem instance allocated in <code>Interface_GUIListItem::create</code> and each <code>CListeItem</code>instance tries to delete it in its dtor => double free!

AlwinEsch replied:
Have looked:

Currently (on old) is this used:
```cpp
CFileItemPtr pItem((CFileItem*)item);
```
What do you thing to use it as:
```cpp
CFileItemPtr pItem(make_shared<CFileItem>(item));
```
and copy the data from the first to the second.
And then allow delete on `Interface_GUIListItem::destroy`?

Is for me also not the best (for cases where addon want to have independent list together) but more safe.

ksooo commented on xbmc/windows/GUIMediaWindow.cpp:
```diff
@@ -565,7 +567,9 @@ void CGUIMediaWindow::ClearFileItems()
 
-// \brief Sorts Fileitems based on the sort method and sort oder provided by guiViewState
+/*!
+ * \brief Sorts Fileitems based on the sort method and sort oder provided by guiViewState
```
oder -> order

AlwinEsch replied:
Becomes changes, also do I move this on GUIMediaWindow.cpp to a separate request

ksooo reviewed (changes requested):
Needs some more love, imo.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -0,0 +1,665 @@
+{
+  // Hook Over calling  CGUIMediaWindow::OnClick(iItem) results in it trying to PLAY the file item
+  // which if its not media is BAD and 99 out of 100 times undesireable.
+  return false;
```
This should really investigated :-D (in python "bool WindowXML::OnClick(int iItem) " sine 8 years). I look to them

AlwinEsch replied:
Function can be removed have looked over the code, becomes normally called with https://github.com/xbmc/xbmc/blob/master/xbmc/windows/GUIMediaWindow.cpp#L289 but with override `OnMessage` can it not come to the function here.

AlwinEsch commented:
Thanks for your review! Really good to fix now the old corpses who move from one place to another ;-) .

Since I am really glad to make these changes, to finally get clean.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -0,0 +1,665 @@
+bool CGUIAddonWindowDialog::OnMessage(CGUIMessage& message)
+{
+  if (message.GetMessage() == GUI_MSG_WINDOW_DEINIT)
+    return CGUIWindow::OnMessage(message);
```
Has looked, goes back to the annals since them come in.

Maybe it was related to the old `ShowOverlay` call:
```cpp
bool CGUIAddonWindowDialog::OnMessage(CGUIMessage &message)
{
  if (message.GetMessage() == GUI_MSG_WINDOW_DEINIT)
  {
    CGUIWindow *pWindow = g_windowManager.GetWindow(g_windowManager.GetActiveWindow());
    if (pWindow)
      g_windowManager.ShowOverlay(pWindow->GetOverlayState());
    return CGUIWindow::OnMessage(message);
  }
  return CGUIAddonWindow::OnMessage(message);
}
```

I check and remove it complete, if not needed.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -0,0 +1,665 @@
+    g_windowManager.Remove(pAddonWindow->GetID());
+  }
+  delete pAddonWindow;
+  Interface_GUIGeneral::unlock();
```
Fixed

ksooo commented:
Ready to merge, imo.

ksooo commented:
jenkins build this please

AlwinEsch committed:
[addons] next binary addon gui interface step

This add the minimum parts to create a window on addon. This still
mostly equal with the Python addon window parts.

During test has this worked, but my experience for GUI things is not the
best and want a feedback.

AlwinEsch commented:
jenkins build this please
